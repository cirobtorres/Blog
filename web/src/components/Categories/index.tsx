import Link from "next/link";
import {
  Tooltip,
  TooltipContent,
  TooltipProvider,
  TooltipTrigger,
} from "../Shadcnui/tooltip";

const Categories = ({
  category,
  subCategories,
  tags,
}: {
  category: Category;
  subCategories?: SubCategory[];
  tags?: Tag[];
}) => {
  return (
    <section className="mb-10">
      <Category category={category} />
      {subCategories && subCategories.length > 0 && (
        <Subcategories subCategories={subCategories} />
      )}
      {tags && tags.length > 0 && <Tags tags={tags} />}
    </section>
  );
};

const Category = ({ category }: { category: Category }) => {
  return (
    <article
      className="max-w-screen-2xl mx-auto mb-12"
      role="region"
      data-testid="article-category"
      aria-label={`Categoria: ${category.name}`}
    >
      <div className="flex justify-center flex-wrap gap-6 col-start-2 max-lg:col-start-1 px-4">
        <div className="w-full grid grid-cols-[1fr_auto_1fr] max-[800px]:grid-cols-1 items-center">
          <hr className="w-full h-[1px] bg-blog-border max-[800px]:hidden" />
          <div className="px-4">
            <h2
              id={`category-title-${category.documentId}`}
              className="text-[2.5rem] font-extrabold text-center max-[800px]:text-[2rem]"
            >
              {category.name}
            </h2>
          </div>
          <hr className="w-full h-[1px] bg-blog-border max-[800px]:hidden" />
        </div>
      </div>
    </article>
  );
};

const Subcategories = ({ subCategories }: { subCategories: SubCategory[] }) => {
  return (
    <article
      data-testid="article-subcategories"
      aria-label="Tecnologias"
      className="max-w-screen-2xl mx-auto mb-10"
    >
      <ul
        data-testid="article-subcategories-list"
        role="list"
        aria-label="Lista das tecnologias abordadas neste artigo"
        className="max-w-[500px] mx-auto flex flex-wrap gap-6 justify-center items-center px-4"
      >
        {subCategories.map((subCategory) => (
          <li
            key={subCategory.documentId}
            data-testid={subCategory.documentId}
            role="listitem"
            className="flex items-center gap-2 h-6"
          >
            <div
              aria-hidden="true" // Decorative element
              dangerouslySetInnerHTML={{ __html: subCategory.svg }}
            />
            <TooltipProvider>
              <Tooltip>
                <TooltipTrigger asChild>
                  <Link
                    href={subCategory.link}
                    target="_blank"
                    rel="noopener noreferrer"
                    aria-label={`Abrir nova aba com site oficial de ${subCategory.name}`}
                    className="text-blog-foreground-readable hover:text-blog-foreground-readable-hover rounded focus-within:outline focus-within:outline-2 focus-within:outline-blog-foreground-readable-hover"
                  >
                    {subCategory.name}
                  </Link>
                </TooltipTrigger>
                <TooltipContent>
                  <p>{subCategory.link}</p>
                </TooltipContent>
              </Tooltip>
            </TooltipProvider>
          </li>
        ))}
      </ul>
    </article>
  );
};

const Tags = ({ tags }: { tags: Tag[] }) => {
  return (
    <article data-testid="article-tags" aria-label="Tags" className="mb-10">
      <div className="max-w-screen-lg mx-auto">
        <ul
          role="list"
          data-testid="article-tags-list"
          aria-label="Lista de tags"
          className="flex justify-center flex-wrap gap-6 blog-heading col-start-2 max-lg:col-start-1 px-40 max-[800px]:px-4"
        >
          {tags.map((tag, index) => (
            <li
              data-testid={tag.documentId}
              key={tag.documentId}
              role="listitem"
              aria-labelledby={`Tag-${index + 1}`}
              className="flex items-center gap-2 h-6"
            >
              <span className="text-[#808080] italic">{tag.name}</span>
            </li>
          ))}
        </ul>
      </div>
    </article>
  );
};

export default Categories;
