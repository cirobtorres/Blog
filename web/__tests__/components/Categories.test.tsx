import Categories from "@/components/Categories";
import { render, screen } from "@testing-library/react";

describe("Categories", () => {
  const categoryMocked = { documentId: "123", name: "Ciência da Computação" };

  const subcategoriesMocked = [
    {
      documentId: "123",
      name: "Next.js",
      link: "https://nextjs.org/",
      svg: '<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 180 180" width="18"><mask height="180" id=":r8:mask0_408_134" maskUnits="userSpaceOnUse" width="180" x="0" y="0" style="mask-type: alpha;"><circle cx="90" cy="90" fill="black" r="90"></circle></mask><g mask="url(#:r8:mask0_408_134)"><circle cx="90" cy="90" data-circle="true" fill="black" r="90"></circle><path d="M149.508 157.52L69.142 54H54V125.97H66.1136V69.3836L139.999 164.845C143.333 162.614 146.509 160.165 149.508 157.52Z" fill="url(#:r8:paint0_linear_408_134)"></path><rect fill="url(#:r8:paint1_linear_408_134)" height="72" width="12" x="115" y="54"></rect></g><defs><linearGradient gradientUnits="userSpaceOnUse" id=":r8:paint0_linear_408_134" x1="109" x2="144.5" y1="116.5" y2="160.5"><stop stop-color="white"></stop><stop offset="1" stop-color="white" stop-opacity="0"></stop></linearGradient><linearGradient gradientUnits="userSpaceOnUse" id=":r8:paint1_linear_408_134" x1="121" x2="120.799" y1="54" y2="106.875"><stop stop-color="white"></stop><stop offset="1" stop-color="white" stop-opacity="0"></stop></linearGradient></defs></svg>',
    },
  ];

  const tagsMocked = [
    {
      documentId: "8",
      name: "Programação",
    },
    {
      documentId: "11",
      name: "linguagem de Programação",
    },
    {
      documentId: "19",
      name: "Banco de Dados",
    },
    {
      documentId: "33",
      name: "Diagrama UML",
    },
    {
      documentId: "35",
      name: "Programação",
    },
  ];

  test("Categories component matches the snapshot", () => {
    const { asFragment } = render(
      <Categories
        category={categoryMocked}
        subCategories={subcategoriesMocked}
        tags={tagsMocked}
      />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // describe("Category", () => {
  //   test("", () => {});
  // });

  // describe("Subcategories", () => {
  //   test("", () => {});
  // });

  describe("Tags", () => {
    test("Tags to not be in the document if tags article has no tag", () => {
      render(
        <Categories
          category={categoryMocked}
          subCategories={subcategoriesMocked}
          tags={[]}
        />
      );

      const tags = screen.queryByTestId("article-tags");
      expect(tags).toBeNull();
      expect(tags).not.toBeInTheDocument();
    });

    test("Tags to be in the document if tags article has tags", () => {
      render(
        <Categories
          category={categoryMocked}
          subCategories={subcategoriesMocked}
          tags={tagsMocked}
        />
      );

      const tags = screen.queryByTestId("article-tags");
      expect(tags).toBeInTheDocument();

      tagsMocked.forEach((tag) => {
        const listItem = screen.getByTestId(tag.documentId);
        expect(listItem).toBeInTheDocument();
      });
    });

    test("Tags accessibility attributes are rendered", () => {
      render(
        <Categories
          category={categoryMocked}
          subCategories={subcategoriesMocked}
          tags={tagsMocked}
        />
      );

      const tagsList = screen.queryByTestId("article-tags-list");
      expect(tagsList).toHaveAttribute("role", "list");
      expect(tagsList).toHaveAttribute("aria-labelledby", "tags-list");

      const listItems = tagsList?.getElementsByTagName("li");
      expect(listItems?.length).toBe(tagsMocked.length);

      Array.from(listItems || []).forEach((listItem, index) => {
        const tag = tagsMocked[index];
        expect(listItem).toHaveAttribute("data-testid", tag.documentId);
        expect(listItem).toHaveAttribute("role", "listitem");
      });
    });
  });
});
