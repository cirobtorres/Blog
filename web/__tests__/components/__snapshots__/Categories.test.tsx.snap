// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories Categories component matches the snapshot 1`] = `
<DocumentFragment>
  <section
    class="mb-10"
  >
    <article
      aria-label="Categoria: Ciência da Computação"
      class="max-w-screen-2xl mx-auto mb-12"
      data-testid="article-category"
      role="region"
    >
      <div
        class="flex justify-center flex-wrap gap-6 col-start-2 max-lg:col-start-1 px-4"
      >
        <div
          class="w-full grid grid-cols-[1fr_auto_1fr] max-[800px]:grid-cols-1 items-center"
        >
          <hr
            class="w-full h-[1px] bg-blog-border max-[800px]:hidden"
          />
          <div
            class="px-4"
          >
            <h2
              class="text-[2.5rem] font-extrabold text-center max-[800px]:text-[2rem]"
              id="category-title-123"
            >
              Ciência da Computação
            </h2>
          </div>
          <hr
            class="w-full h-[1px] bg-blog-border max-[800px]:hidden"
          />
        </div>
      </div>
    </article>
    <article
      aria-label="Tecnologias"
      class="max-w-screen-2xl mx-auto mb-10"
      data-testid="article-subcategories"
    >
      <ul
        aria-label="Lista das tecnologias abordadas neste artigo"
        class="max-w-[500px] mx-auto flex flex-wrap gap-6 justify-center items-center px-4"
        data-testid="article-subcategories-list"
        role="list"
      >
        <li
          class="flex items-center gap-2 h-6"
          data-testid="14"
          role="listitem"
        >
          <div
            aria-hidden="true"
          />
          <a
            aria-label="Abrir nova aba com site oficial de Next.js"
            class="transition-all duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover"
            data-state="closed"
            href="https://nextjs.org/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Next.js
          </a>
        </li>
        <li
          class="flex items-center gap-2 h-6"
          data-testid="16"
          role="listitem"
        >
          <div
            aria-hidden="true"
          />
          <a
            aria-label="Abrir nova aba com site oficial de Nest.js"
            class="transition-all duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover"
            data-state="closed"
            href="https://nestjs.com/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Nest.js
          </a>
        </li>
      </ul>
    </article>
    <article
      aria-label="Tags"
      class="mb-10"
      data-testid="article-tags"
    >
      <div
        class="max-w-screen-lg mx-auto"
      >
        <ul
          aria-label="Lista de tags"
          class="flex justify-center flex-wrap gap-6 blog-heading col-start-2 max-lg:col-start-1 px-40 max-[800px]:px-4"
          data-testid="article-tags-list"
          role="list"
        >
          <li
            aria-labelledby="Tag-1"
            class="flex items-center gap-2 h-6"
            data-testid="8"
            role="listitem"
          >
            <span
              class="text-[#808080] italic"
            >
              Programação
            </span>
          </li>
          <li
            aria-labelledby="Tag-2"
            class="flex items-center gap-2 h-6"
            data-testid="11"
            role="listitem"
          >
            <span
              class="text-[#808080] italic"
            >
              linguagem de Programação
            </span>
          </li>
          <li
            aria-labelledby="Tag-3"
            class="flex items-center gap-2 h-6"
            data-testid="19"
            role="listitem"
          >
            <span
              class="text-[#808080] italic"
            >
              Banco de Dados
            </span>
          </li>
          <li
            aria-labelledby="Tag-4"
            class="flex items-center gap-2 h-6"
            data-testid="33"
            role="listitem"
          >
            <span
              class="text-[#808080] italic"
            >
              Diagrama UML
            </span>
          </li>
          <li
            aria-labelledby="Tag-5"
            class="flex items-center gap-2 h-6"
            data-testid="35"
            role="listitem"
          >
            <span
              class="text-[#808080] italic"
            >
              Programação
            </span>
          </li>
        </ul>
      </div>
    </article>
  </section>
</DocumentFragment>
`;
